### ISSUE TITLE
Add gutter, margin, and padding controls to the columns block

### ISSUE THREAD

**Issue by LevinMedia** [WordPress/gutenberg#24874]
### Limitations with the columns block: 

As it stands, a user is somewhat limited in what they can achieve with the columns block. This is problematic, as it's fundamental building block for constructing a page's layout.  Let's look at an example: the screen shot below is a design I made in Figma. It's a single row with a light yellow background, on top of which sits a simple three column layout. The two outer columns have images, and there's some text and a button in the middle column. 

<img width="1813" alt="Screen Shot 2020-08-27 at 11 56 47 AM" src="https://user-images.githubusercontent.com/4500952/91483340-6a85a480-e85c-11ea-8ad2-ba0ac1674593.png">

The row (parent container) should be the full width of the site. It should have no margin or padding. There are no gutters between the columns. The content in the columns should be aligned to the middle the of row vertically. The columns which contain the images should have no margin or padding. However, the column in the middle should have a healthy amount of padding on the left and right sides to keep the headline more condensed and in the center of the column. 

This is the best I could do using the the blocks in core:

<img width="2559" alt="Screen Shot 2020-08-27 at 11 36 46 AM" src="https://user-images.githubusercontent.com/4500952/91481672-a10df000-e859-11ea-9506-cde68e571059.png">

- The parent container (The columns block) has padding automatically applied to it. There are no controls to change or remove it. 
- A user is unable to add or remove margin to the parent block as well.  This becomes problematic if you wish to stack multiple rows directly on top of each other, with no space in-between them.
- There's no ability to control the gutters between the columns, or remove them. 
- There's no ability to control the margin or padding that's applied to the individual columns. 

### Solutions:

- Remove the padding that's automatically applied to the columns block
- Add margin and padding controls to the columns block (parent block)
- Add gutter controls to the columns block (parent block)
- Add margin and padding controls to individual columns blocks (child blocks)

**One small detail to keep in mind:** at the moment, when a background color is added to many types of blocks (paragraph, headline, column etc) padding is automatically added as well. This behavior should either be eliminated, **_or_** the auto padding value should appear in the new padding control component, there by allowing the user to refine it further, or remove it.

**Controls for the "columns block" (parent) and "column block" (child) would look something like this:**

<img width="470" alt="Screen Shot 2020-08-27 at 12 49 29 PM" src="https://user-images.githubusercontent.com/4500952/91488245-de777b00-e863-11ea-8979-6ee45a3fd220.png">

both of these are using the new [box control](https://wordpress.github.io/gutenberg/?path=/story/components-boxcontrol--default) component, which was recently included on the cover block.

### Bigger picture, and out of scope for this particular issue:

Essentially every block should have these margin and padding controls by default.

### Describe alternatives you've considered:

Using Elementor with no prior experience, I was able to re-create this layout in about 10 minutes. Screen shot below:

<img width="2560" alt="Screen Shot 2020-08-27 at 12 15 21 PM" src="https://user-images.githubusercontent.com/4500952/91485113-28119700-e85f-11ea-853f-e6634c2336cc.png">

Elementor uses the same principle which I've described in the solutions above. There's no margin or padding applied to the parent container by default. The parent container has controls for the gutters. Each of the elements (Parent block, and child column blocks) have their own margin and padding controls.

cc @jameskoster

**Comment by jameskoster** [comment]
> One small detail to keep in mind: at the moment, when a background color is added numerous block (paragraph, headline, column etc) padding is automatically added to it. This behavior should either be eliminated, or the auto padding value should appear in the new padding control component, there by allowing the user to refine it further, or remove it.

One of the key considerations here is what happens to existing instances of the block upon update. To avoid disruption on the frontend, wherever folks are currently using the background color option with the static auto-padding applied, that auto-padding value should probably be migrated to the new dedicated padding control.

**Comment by richtabor** [comment]
I've found margin to be very tricky to implement in blocks. Padding is great, but margin is difficult because it interacts with the space _outside_ of the block, whereas padding is inside. When interacting outside, it becomes tricky to emulate one-to-one with the front-end, and creates confusion on how to manipulate that spacing (as not every block has margin controls - rightfully so). I lean towards using Spacer blocks for "margin" control.

Gutter is tricky for mobile implementation, but it's not impossible. We (CoBlocks) figured out some creative solutions for gutters in a few blocks; including using custom gutter spacing values (with inline CSS vars).

**Comment by aristath** [comment]
Would converting blocks to css-grid with a flex fallback for old browsers be out of the question?
It would solve pretty much all issues regarding gutters, responsive etc.

**Comment by lifeact** [comment]
Add gutter, margin, and padding controls to the any block

**Comment by andrewserong** [comment]
I believe this one is now resolved, as of the following PRs:

* Columns block gap / spacing support and vertical margin support added in: https://github.com/WordPress/gutenberg/pull/34630
* Columns padding support added in: https://github.com/WordPress/gutenberg/pull/35701

Themes can opt in to the above controls by setting the following spacing settings in `theme.json`:

```
		"spacing": {
			"blockGap": true,
			"margin": true,
			"padding": true,
```

I'll close this issue out now.

**Comment by ethanclevenger91** [comment]
@andrewserong  so far as I can tell it's:

```
"spacing": {
	"customMargin": true,
	"customPadding": true
},
```

Right?

**Comment by andrewserong** [comment]
@ethanclevenger91 yes, up until yesterday (and in the v1 of `theme.json`) the margin and padding properties have the `custom` prefix. As of https://github.com/WordPress/gutenberg/pull/36155 in the v2 of `theme.json` the `custom` prefix has been removed. It looks like the `v1` naming will still continue to work though 🙂

**Comment by Nic787** [comment]
Is there a way to enable gap support via functions.php? I don't use theme.json

In my functions file, I have enabled custom spacing, but all the new spacing options are not showing, so I don't know if it was meant to be in 5.9 or not.

Thank you

**Comment by andrewserong** [comment]
> Is there a way to enable gap support via functions.php? I don't use theme.json

@Nic787 I don't believe so. The new spacing controls are designed to work with the `theme.json` approach, and I think the only controls that are switched on via `add_theme_support( 'custom-spacing' );` are those for [backwards compatibility](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#backward-compatibility-with-add_theme_support) (in this case, padding — [here's where the check happens](https://github.com/wordpress/gutenberg/blob/trunk/packages/block-editor/src/components/use-setting/index.js#L51) in the Gutenberg code).

It should be possible to add [a minimal `theme.json` file](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/) to an existing theme in 5.9 if you'd like to opt in to the new controls (the [appearanceTools](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls) setting will opt-in to most of them). From memory, I think part of the decision to guard features behind opting in via `theme.json` was to ensure that existing themes don't unintentionally expose new controls. For example, the block gap support depends on the layout support, which renders additional styling that could conflict with some classic themes.

**Comment by Nic787** [comment]
@andrewserong I did create a short theme.json file.
```
{
    "version": 2,
    "settings": {
        "spacing": {
			"blockGap": true,
			"customPadding": true,
			"customMargin": true,
			"units": [
				"%",
				"px",
				"em",
				"rem",
				"vh",
				"vw"
			]
		}
    }
}
```
However, it doesn't work and it also remove all the default style of Gutenberg block editor.

![image](https://user-images.githubusercontent.com/11185348/153721711-4e1dbea4-e768-4f38-ad75-c2d24e5f4064.png)

Also, how can I bring back default Wordpress style in the editor?

**Comment by carolinan** [comment]
@Nic787 Version 2 does not use "customPadding": true, "customMargin": true, so this creates an error in theme.json.
Use "padding": true, "margin": true instead.

Please use this reference:
https://developer.wordpress.org/block-editor/reference-guides/theme-json-reference/theme-json-living/

**Comment by Nic787** [comment]
> 

My bad, I changed it after I wrote it.

For now I use:

```
{
    "version": 2,
    "settings": {
        "appearanceTools": true,
        "layout": {
			"contentSize": true,
			"wideSize": true
		},
        "spacing": {
                        "blockGap": true,
			"units": [
				"%",
				"px",
				"em",
				"rem",
				"vh",
				"vw"
			]
		}
    }
}
```

Why does it look like that?
![image](https://user-images.githubusercontent.com/11185348/153762827-a846264a-ae20-4345-a195-4cb1f2d80337.png)
Instead of that?
![image](https://user-images.githubusercontent.com/11185348/153762870-413263bd-74f7-42a8-ba2d-837582d86dc6.png)

But still unable to get the block gap functionality working.

EDIT: I was able to fix the editor style. In fact, I thought that the layout option was affecting my site style, but looks like it's affecting the editor layout instead. So I fixed that.

But why my block gaps not working? Do I have to add something else on my website? Does it just work for gallery block or am I supposed to have it on columns too?
